src/System.cpp

```diff
@@ -421,10 +421,15 @@ void CSystem::ResetMem(unsigned int membits) {
 /**
  * Register a device.
  **/
-int CSystem::RegisterComponent(CSystemComponent *component) {
+void CSystem::RegisterComponent(CSystemComponent *component) {
   acComponents[iNumComponents] = component;
   iNumComponents++;
-  return 0;
+}
+
+void CSystem::UnregisterComponent(CSystemComponent *component) {
+  if (acComponents[iNumComponents] == component)
```

Comment:
I think comparing by memory address is less likely to cause race condition when components are destructed i another order. A counter would have to be atomic, and a for loop to compare all items would have the same crash behavior.

Catching the exception at creation and then handling it would be another solution but that would also require some refactoring of the component registration (it happens so early in the base class, it be better if it was registering after constructing for this bug).

I could try to just not set it to a nullptr because it will be deleted right after the call to stop threads? Or try how a std::atomic counter would perform? What has your preference?

Reply:
You are right - I missed that the component number couldn't be read from an uninitialized component. I take my comment back, you can keep the code as it is.

Reply:
> I could try to just not set it to a nullptr because it will be deleted right after the call to stop threads? Or try how a std::atomic counter would perform? What has your preference?

I might be missing something, but doesn't the if before the nullptr setting check the wrong index? `iNumComponents` is increased in `CSystem::RegisterComponent`, so it should point to the next (empty) element, not to the last one.

Reply:
Yes you are right, it checks the wrong component. The entire check can be removed, I'll do that right away.

Reply:
That's done, now it just decrements the counter.